lib/complete-provider.coffee

```diff
@@ -53,11 +58,19 @@ class CompleteProvider extends Provider
     'then', 'where'
   ]
 
-  initialize: (@editorView, @manager) ->
+  constructor: (@editor, @manager) ->
+    if (@editor.haskellCompletionProvider && @editor.haskellCompletionProvider != this)
+      # shouldn't happen, but log just in case
+      console.log "weird: editor already has a haskell completion provider"
+      @editor.haskellCompletionProvider.dispose()
```

Comment:
It's a little trickier to use WeakMap here, but basically, since PluginManager is a singleton, you might maintain it there and use some sort of accessor like e.g.
```coffeescript
providerForEditor: (editor) =>
    @editorMap.get(editor)
```